designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,627 @@
+
+```js
+exports.config = (context) => {
+    const myConfig = {
+        files: ["*.js"],
+        ruledefs: {
+            custom: ruleLoader.loadFromDirectory("./custom-rules")
+        },
+        rules: {
+            "custom/my-rule": "error"
+        }
+    };
+
+    if (context.hasRule("some-new-rule")) {
+        myConfig.rules["some-new-rule"] = ["error"];
+    }
+
+    return myConfig;
+};
+```
+
+### Configuration Location Resolution
+
+When ESLint is executed, the following steps are taken to find the `eslint.config.js` file to use:
+
+1. If the `-c` flag is used then the specified configuration file is used. There is no further search performed.
+1. Otherwise:
+    a. Look for `eslint.config.js` in the current working directory. If found, stop searching and use that file.
+    b. If not found, search up the directory hierarchy looking for `eslint.config.js`.
+    c. If a `eslint.config.js` file is found at any point, stop searching and use that file.
+    d. If `/` is reached without finding `eslint.config.js`, then stop searching and output a "no configuration found" error.
+
+This approach will allow running ESLint from within a subdirectory of a project and get the same result as when ESLint is run from the project's root directory (the one where `eslint.config.js` is found).
+
+Some of the key differences from the way ESLint's configuration resolution works today are:
+
+1. There is no automatic search for `eslint.config.js` in the user's home directory. Users wanting this functionality can either pass a home directory file using `-c` or manually read in that file from their `eslint.config.js` file.
+1. Once a `eslint.config.js` file is found, there is no more searching for any further config files.
+1. There is no automatic merging of config files using either `extends` or `overrides`.
+1. When `-c` is passed on the command line, there is no search performed.
+
+### File Pattern Resolution
+
+Because there are file patterns included in `eslint.config.js`, this requires a change to how ESLint decides which files to lint. The process for determining which files to lint is:
+
+1. When a filename is passed directly (such as `eslint foo.js`):
+    1. ESLint checks to see if there is one or more configs where the `files` pattern matches the file that was passed in and does not match the `ignores` pattern. The pattern is evaluated by prepending the directory in which `eslint.config.js` was found to each pattern in `files`. All configs that match `files` and not `ignores` are merged (with the last matching config taking precedence over others). If no config is found, then the file is ignored with an appropriate warning.
```

Comment:
How does this affect default ignore patterns such as `node_modules`? I'm assuming we would still ignore `node_modules` by default when running `eslint .`, but would `node_modules/foo.js` still be ignored if it was explicitly listed in a `files` pattern?

Reply:
That's a good question. I was assuming that people would manually specify to ignore `node_modules`, but we probably do want to ignore that by default as we do now. I'll add in a note about that.